global_helpers/panther_lookuptable_helpers.py

```diff
@@ -0,0 +1,16 @@
+from panther_base_helpers import deep_get
+
+
+# pylint: disable=too-few-public-methods
+class LookupTableMatches:
+    def _register(self, event, lookuptable_name: str):
+        # pylint: disable=attribute-defined-outside-init
+        self.lut_matches = deep_get(event, "p_enrichment", lookuptable_name)
+
+    def _lookup(self, match_field: str, *keys) -> list or str:
+        match = deep_get(self.lut_matches, match_field)
+        if not match:
+            return None
+        if isinstance(match, list):
+            return [deep_get(match_value, *keys) if match_value else None for match_value in match]
```

Comment:
One thing to consider with adding list support is the ordering of the values in `p_enrichment`. The way to do it is to make sure that the list in `p_enrichment` matches the input exactly, i.e. `in[X] = enriched[X]` 

For example:
```
in = [
 'missing', 
 'foo',
 'foo', # repeated
]
enriched = [
 '', # missing value is str for proper typing 
 'bar', 
 'bar', # repeated val is listed twice
```

We should make sure this is how the detection engine behaves already.

Reply:
Are you suggesting we "pad" the array with an empty `{}` if there is no match to have alignment?

Reply:
I don't think the detections engine works that way ... I think you get the matching subset

Reply:
exactly, otherwise it is not possible to know which enrichment maps to which input in the list

Reply:
I think we need to modify the detections engine then

Reply:
I just looked at the code in the detections engine, it just collects the matches so correspondence is lost unless all match.

Here is my proposal for the small change:
- Non-matching "slots" get a `None` ... this maintains the 1-1 correspondence
- Results are  only returned if at least 1 "slot" matched (same logic as now really).

Thoughts?

Reply:
Also any function that returns None will need an updated type annotation: `Union[..., None]`

Reply:
I think we solved this issue with this: https://github.com/panther-labs/panther-enterprise/pull/13012